lib/execution_segment.go

```diff
@@ -394,13 +395,151 @@ func (ess ExecutionSegmentSequence) String() string {
+	// this has the consequence that if this is ran without sequence,
+	// but with segments: 0:1/3 and 1/3:2/3 it will get the same results instead
+	// of 1/3:2/3 to get start=1 and offset={3} it will get as 0:1/3 will start=0 and offsets={3}
+	// if the above behaviour is desired this will definitely need to be outside of this function.
+	if ess == nil || len(*ess) == 0 {
+		copyESS = []*ExecutionSegment{segment}
+	} else {
+		copyESS = append([]*ExecutionSegment{}, *ess...) // copy the original sequence
+	}
+	var wrapper = newWrapper(copyESS)
+
+	var segmentIndex = wrapper.indexOf(segment)
+	if segmentIndex == -1 {
+		return -1, nil, -1, fmt.Errorf("missing segment %s inside segment sequence %s", segment, ess)
+	}
+	start, offsets := wrapper.strippedOffsetsFor(segmentIndex)
+	return start, offsets, wrapper.lcd, nil
+}
+
+// This is only needed in order to sort all three at the same time
+type sortInterfaceWrapper struct { // TODO: rename ?
+	ess        ExecutionSegmentSequence
+	numerators []int64
+	lcd        int64
+}
+
+func newWrapper(ess ExecutionSegmentSequence) sortInterfaceWrapper {
+	var result = sortInterfaceWrapper{
+		ess:        ess,
+		numerators: make([]int64, len(ess)),
+		lcd:        ess.lcd(),
+	}
+
+	for i := range ess {
+		result.numerators[i] = ess[i].length.Num().Int64() * (result.lcd / ess[i].length.Denom().Int64())
+	}
+
+	sort.Stable(result)
+	return result
+}
+
+func (e sortInterfaceWrapper) indexOf(segment *ExecutionSegment) int {
+	for i, seg := range e.ess {
+		if seg.Equal(segment) {
+			return i
+		}
+	}
+
+	return -1
+}
+
+func (e sortInterfaceWrapper) strippedOffsetsFor(segmentIndex int) (int64, []int64) {
+	var offsets = make([]int64, 0, e.numerators[segmentIndex]+1)
+	var chosenCounts = make([]int64, len(e.ess))
+	// Here instead of calculating steps which need to be big.Rat, we use the fact that
+	// the steps are always the length of the segment inverted which also is lcd/numerator
+	// So instead of creating and adding up big.Rat we just multiply the step by the amount
+	// of times given segment has been chosen which is count * lcd / numerator and use that
+	// this both saves on a lot of big.Rat allocations and also on a lot of unneeded calculations
+	// with them.
+
+	for i := int64(0); i < e.lcd; i++ {
+		for index, chosenCount := range chosenCounts {
+			num := chosenCount * e.lcd
+			denom := e.numerators[index]
+			if i > num/denom || (i == num/denom && num%denom == 0) {
+				chosenCounts[index]++
+				if index == segmentIndex {
+					prev := int64(0)
+					if len(offsets) > 0 {
+						prev = offsets[len(offsets)-1]
+					}
+					offsets = append(offsets, i-prev)
+					if int64(len(offsets)) == e.numerators[index] {
+						offsets = append(offsets, offsets[0]+e.lcd-i)
+						return offsets[0], offsets[1:]
+					}
+				}
+				break
+			}
+		}
+	}
+
+	// TODO return some error if we get to here
+	return offsets[0], offsets[1:]
```

Comment:
With the current implementation, this seems like it should be a `panic()`, since it shouldn't be reached unless there's a bug, right?

Reply:
yeah ... Although I still wonder if I shouldn't implement this in some way that you can calculate the offsets for all segments. In that case, maybe the implementation will be such that this will be avoided altogether ... I didn't manage to think of a good way to use the same code at the time and didn't want to copy-paste the code or have to calculate all the offsets when I only need one :D .
But yeah in the current situation this shouldn't happen and it is not covered by tests ... as it never happens :D

Reply:
See my other comment - https://github.com/loadimpact/k6/pull/1323#discussion_r383874754
If, instead of returning in the middle of the function, you accumulate values with a lambda, it should be both more readable and there wouldn't be unreachable code

Reply:
you and your lambdas .... will try it out :D